### PR TITLE
AK: Math portability improvements

### PR DESCRIPTION
**Pull request by Hendiadyoin1**
Makes AK/Math more portable, and fixes M1 Lagom buid

### CODE REVIEW

**Comment by Hendiadyoin1** [comment]
I'll rebase this this week, it should still be relevant

**Comment by creator1creeper1** [comment]
`AK/bits/Math/libm_fallback.h` is a file in a weird location with some weird contents.

File: AK/Math.h [hunk 1]
```diff
@@ -304,7 +305,7 @@ CONSTEXPR T log2(T x)
 template<Integral T>
 CONSTEXPR T log2(T x)
 {
-    return x ? 8 * sizeof(T) - count_leading_zeroes(static_cast<MakeUnsigned<T>>(x)) : 0;
+    return x ? (8 * sizeof(T) - count_leading_zeroes((AK::Detail::MakeUnsigned<T>)x) - 1) : NumericLimits<T>::min();
```
Comment (bgianfo) [inline AK/Math.h:95]: Why did we need to remove the static cast? Seems unrelated?

File: AK/bits/Math/libm_fallback.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ * Copyright (c) 2021, Leon Albrecht <leon2002.la@gmail.com>.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Concepts.h>
+#include <AK/Types.h>
+#include <math.h>
```
Comment (bgianfo) [inline AK/bits/Math/libm_fallback.h:11]: We can’t include math.h in Kernel mode?

File: AK/bits/Math/libm_fallback.h [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+#ifdef __clang__
+#    define CONSTEXPR ALWAYS_INLINE
```
Comment (bgianfo) [inline AK/bits/Math/libm_fallback.h:14]: These defines are now defined multiple places, can we consolidate?

File: AK/Math.h [hunk 4]
```diff
@@ -36,6 +36,12 @@ template<size_t value>
 constexpr size_t product_odd() { return value * product_odd<value - 2>(); }
 }
 
+#ifdef __clang__
+#    define CONSTEXPR ALWAYS_INLINE
```
Comment (bgianfo) [inline AK/Math.h:25]: I don’t think naming this CONSTEXPR when it won’t be in some compilers isa good idea, please invent a new name.

File: AK/Math.h [hunk 5]
```diff
@@ -21,243 +21,54 @@ constexpr T Pi = 3.141592653589793238462643383279502884L;
 template<FloatingPoint T>
 constexpr T E = 2.718281828459045235360287471352662498L;
 
-namespace Details {
-template<size_t>
-constexpr size_t product_even();
-template<>
-constexpr size_t product_even<2>() { return 2; }
-template<size_t value>
-constexpr size_t product_even() { return value * product_even<value - 2>(); }
```
Comment (bgianfo) [inline AK/Math.h:29]: How are these architecture specific?

**Review by bgianfo — CHANGES_REQUESTED** [review]

File: AK/Math.h (hunk 1 above)
Comment (Hendiadyoin1) [reply AK/Math.h:95]: oops

File: AK/bits/Math/libm_fallback.h (hunk 2 above)
Comment (Hendiadyoin1) [reply AK/bits/Math/libm_fallback.h:11]: Math.h should only be used in Userland
and the libm_fallback is for lagom on non-x86 platforms only, otherwise it would call back to itself.

File: AK/Math.h (hunk 4 above)
Comment (Hendiadyoin1) [reply AK/Math.h:25]: Well I am even thinking of just marking these as just "inline"/"static inline",
because some of the implementations are gonna be big enough, to be not ideal to inline in the not constant evaluated case...
Sadly we cant mutate constexpr-ness in specializations, can we do so for __attribute__'s though?

File: AK/Math.h (hunk 5 above)
Comment (Hendiadyoin1) [reply AK/Math.h:29]: I moved them because they are/should only used by the x86 implementation.

File: AK/bits/Math/libm_fallback.h (hunk 2 above)
Comment (Hendiadyoin1) [reply AK/bits/Math/libm_fallback.h:13]: (also I am bad at naming....)

File: AK/Math.h (hunk 4 above)
Comment (Hendiadyoin1) [reply AK/Math.h]: also I am bad at naming...

Comment (BertalanD) [reply AK/Math.h]: Please drop this change. There is no real benefit to forcing these to be inlined, and just makes the code more confusing. Even if these never produce constant expressions on Clang, `constexpr` implies `inline` regardless.

**Review by BertalanD — CHANGES_REQUESTED** [review]

File: AK/bits/Math/libm_fallback.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+template<FloatingPoint T>
+CONSTEXPR T asinh(T x) { return ::asinh(x); }
+template<FloatingPoint T>
+CONSTEXPR T acosh(T x) { return ::acosh(x); }
+template<FloatingPoint T>
+CONSTEXPR T atanh(T x) { return ::atanh(x); }
+}
+
+template<FloatingPoint T>
+T pow(T x, T y) { return ::pow(x, y); };
```
Comment (mnlrsn) [inline AK/bits/Math/libm_fallback.h:81]: Error message received:
serenity/Meta/Lagom/../../AK/bits/Math/libm_fallback.h:81:3: error: redeclaration 'T AK::pow(T, T)' differs in 'constexpr' from previous declaration

Seems to be fixed by adding CONSTEXPR also for T pow(T x, T y) .

File: AK/Math.h [hunk 7]
```diff
@@ -8,249 +8,55 @@
...
-    do {
-        asm(
-            "fprem\n"
-            "fnstsw %%ax\n"
-            : "+t"(x), "=a"(fpu_status)
-            : "u"(y));
-    } while (fpu_status & 0x400);
-    return x;
-}
+CONSTEXPR T fmod(T x, T y);
```
Comment (BertalanD) [inline AK/Math.h:20]: Declaring a function `constexpr` without giving it an actual definition doesn't accomplish anything.

If the definition comes from the architecture-specific headers, why bother declaring them here as well?

Or we can just forward to `cmath`'s `std` namespaced functions that already support overloading.

File: AK/bits/Math/libm_fallback.h [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+CONSTEXPR T log10(T x) { return ::log10(x); }
+template<FloatingPoint T>
+CONSTEXPR T exp(T exponent) { return ::exp(exponent); }
+template<FloatingPoint T>
+CONSTEXPR T exp2(T exponent) { return ::exp2(exponent); }
+}
+
+namespace Hyperbolic {
+template<FloatingPoint T>
+CONSTEXPR T sinh(T x) { return ::sinh(x); }
```
Comment (BertalanD) [inline AK/bits/Math/libm_fallback.h:67]: Are these ever going to be constant expressions? 🤔

We should probably use function overloading to use the right LibM call based on the type of the argument.

**Review by BertalanD — CHANGES_REQUESTED** [review]

File: AK/Math.h (hunk 7 above)
Comment (Hendiadyoin1) [reply AK/Math.h]: To have a easy to find list of all functions supported/needed to be implemented

File: AK/bits/Math/libm_fallback.h (hunk 8 above)
Comment (Hendiadyoin1) [reply AK/bits/Math/libm_fallback.h]: I'll change it out with cmath then
I thought normal math had that feature, tooo

File: AK/Math.h (hunk 7 above)
Comment (mnlrsn) [reply AK/Math.h]: > To have a easy to find list of all functions supported/needed to be implemented

Is that not what comments are for? The CONSTEXPR macro _is_ a bit confusing, mostly because the true purpose is not clear. So if the main reason is to flag code that needs to be refactored, then maybe it should be given a second round of thought?

Comment (Hendiadyoin1) [reply AK/Math.h]: The Macro was removed....

File: AK/bits/Math/x86.h [hunk 9]
```diff
@@ -0,0 +1,416 @@
...
+
+template<size_t>
+constexpr size_t product_odd();
+template<>
+constexpr size_t product_odd<1>() { return 1; }
+template<size_t value>
+constexpr size_t product_odd() { return value * product_odd<value - 2>(); }
+}
+
+#define constexpr_STATE(function, args...)        \
```
Comment (mnlrsn) [inline AK/bits/Math/x86.h:34]: Is the owner-casing of "constexpr" in this macro intentional?

Also: The name "CONSTEXPR_STATE" seems historical, but is it still the most expressive name?

Comment (Hendiadyoin1) [reply AK/bits/Math/x86.h:34]: Ohh that was some overeager use of search and replace
And its a macro so it should stay all caps, if you have a better name I can change it, this was what it was in the original implementation

Comment (mnlrsn) [reply AK/bits/Math/x86.h:34]: They say that naming is the hardest problem in computer science :) so I am afraid that I don't have much to offer here... TBH, my biggest gripe with that macro is that it contains a return statement so the execution can return from the method in a way that is a bit hidden from a casual reader of the code. But that is a larger issue, I presume.

File: AK/bits/Math/common.h [hunk 10]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2022, Leon Albrecht <leon2002.la@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
```
Comment (ADKaster) [inline AK/bits/Math/common.h:7]: I'm not sure i'm getting why the path AK/bits/Math is necessary? Seems a flatter structure would do the trick? The bits/ thing seems inherited from glibc. Math would be the only subdirectory of the new `bits` directory. Other files have done like, FooFeature.h and FooFeatureDetails.h.

File: AK/Math.h [hunk 11]
```diff
@@ -428,31 +129,11 @@ using Hyperbolic::sinh;
...
+constexpr T pow(T x, T y);
 }
 
-#undef CONSTEXPR_STATE
-
-}
+#if ARCH(I386) || ARCH(X86_64)
+#    include <AK/bits/Math/x86.h>
+#elif not defined(__serenity__)
+#    include <AK/bits/Math/libm_fallback.h>
```
Comment (ADKaster) [inline AK/Math.h:138]: This file is now implemented as `stl_fallback`, could we update the name?

File: AK/bits/Math/libm_fallback.h [hunk 12]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2021, Leon Albrecht <leon2002.la@gmail.com>.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include "./common.h"
```
Comment (ADKaster) [inline AK/bits/Math/libm_fallback.h:9]: Since we explicitly expect Lagom and possibly non-serenity project owned users to eventually include this file, the relative path "./common.h" include seems potentially dangerous if the project trying to use Lagom has its own "common.h". Can we use the full path here? `<AK/bits/Math/common.h>`.

File: AK/bits/Math/common.h (hunk 10 above)
Comment (ADKaster) [reply AK/bits/Math/common.h:7]: The existence of this file is also confusing. It seems like the only times it's actually used in the arch-specific headers is when someone like HeaderCheck includes the "arch impl" headers directly? Otherwise it's already there from Math.h.

**Review by ADKaster — COMMENTED** [review]
Overall I really like this change and I think it'll really help aarch64 actually become a thing, but while you're resolving conflicts I have a few comments :D.

File: AK/bits/Math/x86.h (hunk 9 above)
Comment (mnlrsn) [reply AK/bits/Math/x86.h:34]: Let me retract my comment about my biggest gripe. That code is now local to x86.h, i.e., part of the special case that is the Intel architecture and hence not a general concern. Please disregard :)

File: Tests/AK/TestMath.cpp [hunk 13]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2021, Leon Albrecht <leon2002.la@gmail.com>.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibTest/TestCase.h>
+
+#include <AK/Math.h>
+#include <AK/NumericLimits.h>
```
Comment (mnlrsn) [inline Tests/AK/TestMath.cpp:10]: Seems to also need

`#include <AK/IntegralMath.h>
`
... at least for `Meta/serenity.sh build lagom` on a Mac

File: AK/bits/Math/common.h (hunk 10 above)
Comment (Hendiadyoin1) [reply AK/bits/Math/common.h]: I put this here, because otherwise clangd will complain about non defined constants etc....

**Comment by Hendiadyoin1** [comment]
Closed in favor of already merged #12648 for now
